lib/dubhash.ts

```diff
@@ -0,0 +1,194 @@
+        format: 1,
+      },
+    ],
+  };
+};
+
+const getDubHashDigests = (
+  fileContents: Buffer,
+  isBinary: boolean,
+): Array<string | null> => {
+  // If the file is binary, you only produce one hash
+  // If the file does not have line endings, you treat it like a binary file
+  // If the LF character is not found, the line endings are UNKNOWN, and we produce only one hash.
+  const lineEndingFormat: LineEndingFormat = detectLineEndingFormat(
+    fileContents,
+  );
+  if (
+    isBinary ||
+    !hasLineEndings(fileContents) ||
+    lineEndingFormat === LineEndingFormat.UNKNOWN
+  ) {
+    return [getSingleDigest(fileContents), null];
+  }
+  // NB: You MUST calculate the digest for the original file and store it, since transformLineEndings
+  // will modify fileContents buffer.
+  const digestForOriginal = getSingleDigest(fileContents);
+  const transformedFile = transformLineEndings(fileContents, lineEndingFormat);
+  return [digestForOriginal, getSingleDigest(transformedFile)];
+};
+
+const hasLineEndings = (fileContents: Buffer): boolean => {
+  const lineEndingBytes = new Uint8Array([0x0a, 0x0d]);
+  for (let readIndex = 0; readIndex < fileContents.length; readIndex++) {
+    if (lineEndingBytes.includes(fileContents[readIndex])) {
+      return true;
+    }
+  }
+  return false;
+};
+
+const detectLineEndingFormat = (fileContents: Buffer): LineEndingFormat => {
+  for (let readIndex = 0; readIndex < fileContents.length; readIndex++) {
+    // If the first character in the file is LF (0x0a), the file is of type LINUX.
+    if (fileContents[readIndex] === 0x0a && readIndex === 0) {
+      return LineEndingFormat.LINUX;
+    }
+    // If the file has a LF character (0x0a), check if the previous character is 0x0d (CR),
+    // then the newline type is WINDOWS
+    if (
+      fileContents[readIndex] === 0x0a &&
+      fileContents[readIndex - 1] === 0x0d
+    ) {
+      return LineEndingFormat.WINDOWS;
+    }
+    if (fileContents[readIndex] === 0x0a) {
+      return LineEndingFormat.LINUX;
+    }
+  }
+  // If the LF character is not found, the line endings are UNKNOWN.
+  return LineEndingFormat.UNKNOWN;
+};
+
+const transformLinuxToWindows = (fileContents: Buffer): Buffer => {
+  // In worst case, file is all 0x0a bytes, so we need to allocate double the memory
+  const transformedFile = Buffer.alloc(2 * fileContents.length);
+  let writeIndex = 0;
+  for (let readIndex = 0; readIndex < fileContents.length; readIndex++) {
+    // if byte is CR (0x0d), do not write byte
+    if (fileContents[readIndex] === 0x0d) {
+      continue;
+    }
+    // if byte is LF and targetLineEndingFormat is WINDOWS, write CR (0x0d) LF (0x0a)
+    // NB1: Check for targetLineEndingFormat === LineEndingFormat.WINDOWS is obviously not necessary here,
+    // NB2: LineEndingFormat.UNKOWN is also not possible here, due to early return in getDubHashDigests
+    if (fileContents[readIndex] === 0x0a) {
```

Comment:
Are there no explicit constants from Node we could use? Or could we define them ourselves? It'd mean less commentary

Reply:
The commentary is intended mainly to help translate between the specification and the typescript code. But I agree a constant would make the code more readable. I couldn't find any node constants to meet our needs, so I will just define them here.